java/src/main/native/src/row_conversion.cu

```diff
@@ -304,20 +334,600 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+ * @param shmem_used_per_block shared memory amount each `block_info` is using
+ * @param block_infos span of `block_info` structs the define the work
+ * @param input_data pointer to raw table data
+ * @param col_sizes array of sizes for each element in a column - one per column
+ * @param col_offsets offset into input data row for each column's start
+ * @param row_offsets offset to a specific row in the output data
+ * @param output_data pointer to output data
+ *
+ */
+__global__ void copy_to_rows(const size_type num_rows, const size_type num_columns,
+                             const size_type shmem_used_per_block,
+                             device_span<const block_info> block_infos, const int8_t **input_data,
+                             const size_type *col_sizes, const size_type *col_offsets,
+                             const size_type *row_offsets, int8_t **output_data) {
+  // We are going to copy the data in two passes.
+  // The first pass copies a chunk of data into shared memory.
+  // The second pass copies that chunk from shared memory out to the final location.
+
+  // Because shared memory is limited we copy a subset of the rows at a time.
+  // This has been broken up for us in the block_info struct, so we don't have
+  // any calculation to do here, but it is important to note.
+
+  constexpr unsigned stages_count = NUM_BLOCKS_PER_KERNEL_LOADED;
+  auto group = cooperative_groups::this_thread_block();
+  extern __shared__ int8_t shared_data[];
+  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_block};
+
+  __shared__ cuda::barrier<cuda::thread_scope_block> block_barrier[NUM_BLOCKS_PER_KERNEL_LOADED];
+  if (group.thread_rank() == 0) {
+    for (int i = 0; i < NUM_BLOCKS_PER_KERNEL_LOADED; ++i) {
+      init(&block_barrier[i], group.size());
+    }
+  }
+
+  group.sync();
+
+  auto const blocks_remaining =
+      std::min(static_cast<uint>(block_infos.size()) - blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS,
+               static_cast<uint>(NUM_BLOCKS_PER_KERNEL_TO_ROWS));
+
+  size_t fetch;
+  size_t subset;
+  for (subset = fetch = 0; subset < blocks_remaining; ++subset) {
+    // Fetch ahead up to stages_count subsets
+    for (; fetch < blocks_remaining && fetch < (subset + stages_count); ++fetch) {
+      auto const fetch_block = block_infos[blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS + fetch];
+      auto const num_fetch_cols = fetch_block.num_cols();
+      auto const num_fetch_rows = fetch_block.num_rows();
+      auto const num_elements_in_block = num_fetch_cols * num_fetch_rows;
+      auto const fetch_block_row_size = fetch_block.get_shared_row_size(col_offsets, col_sizes);
+      auto const starting_column_offset = col_offsets[fetch_block.start_col];
+      auto &fetch_barrier = block_barrier[fetch % NUM_BLOCKS_PER_KERNEL_LOADED];
+
+      // wait for the last use of the memory to be completed
+      if (fetch >= NUM_BLOCKS_PER_KERNEL_LOADED) {
```

Comment:
Under what scenario would this condition be met? It appears that `fetch == 0 || fetch == 1`.

Reply:
The only way is if the constexpr's at the top of the file change to allow each CUDA block to process more data blocks per run. I wanted to allow for play with these values for performance testing.